example/multi_threaded_inference/multi_threaded_inference.cc

```diff
@@ -0,0 +1,351 @@
+               output.end());
+  return output;
+}
+
+std::vector<std::string> LoadSynset(const std::string& synset_file) {
+  std::ifstream fi(synset_file.c_str());
+
+  if (!fi.is_open()) {
+    std::cerr << "Error opening synset file " << synset_file << std::endl;
+    assert(false);
+  }
+
+  std::vector<std::string> output;
+
+  std::string synset, lemma;
+  while (fi >> synset) {
+    getline(fi, lemma);
+    output.push_back(lemma);
+  }
+
+  fi.close();
+
+  return output;
+}
+
+void PrintOutputResult(const float* data, size_t size, const std::vector<std::string>& synset) {
+  if (size != synset.size()) {
+    std::cerr << "Result data and synset size do not match!" << std::endl;
+  }
+
+  float best_accuracy = 0.0;
+  std::size_t best_idx = 0;
+
+  for (std::size_t i = 0; i < size; ++i) {
+    if (data[i] > best_accuracy) {
+      best_accuracy = data[i];
+      best_idx = i;
+    }
+  }
+
+  std::cout << "Best Result: " << trim(synset[best_idx]) << " (id=" << best_idx << ", " <<
+            "accuracy=" << std::setprecision(8) << best_accuracy << ")" << std::endl;
+}
+
+
+// Read Image data into a float array
+void GetImageFile(const std::string &image_file, mx_float *image_data,
+                  int channels, cv::Size resize_size) {
+  // Read all kinds of file into a BGR color 3 channels image
+  cv::Mat im_ori = cv::imread(image_file, cv::IMREAD_COLOR);
+
+  if (im_ori.empty()) {
+    std::cerr << "Can't open the image. Plase check " << image_file << ". \n";
+    assert(false);
+  }
+
+  cv::Mat im;
+  resize(im_ori, im, resize_size);
+
+  int size = im.rows * im.cols * channels;
+
+  mx_float* ptr_image_r = image_data;
+  mx_float* ptr_image_g = image_data + size / 3;
+  mx_float* ptr_image_b = image_data + size / 3 * 2;
+
+  float mean_b, mean_g, mean_r;
+  mean_b = mean_g = mean_r = DEFAULT_MEAN;
+
+  for (int i = 0; i < im.rows; ++i) {
+    auto data = im.ptr<uchar>(i);
+    for (int j = 0; j < im.cols; j++) {
+      if (channels > 1) {
+        *ptr_image_b++ = static_cast<mx_float>(*data++) - mean_b;
+        *ptr_image_g++ = static_cast<mx_float>(*data++) - mean_g;
+      }
+    }
+    *ptr_image_r++ = static_cast<mx_float>(*data++) - mean_r;
+  }
+}
+
+void prepare_input_data(const mxnet::cpp::Shape& shape, const mxnet::cpp::Context& ctx,
+                        int num_threads,
+                        std::vector<mxnet::cpp::NDArray>* data_arr,
+                        bool random_uniform = false) {
+  for (size_t i = 0; i < num_threads; ++i) {
+    data_arr->emplace_back(shape, ctx, false, 0);
+    int begin = i * 100;
+    int end = begin + 100;
+    if (random_uniform) {
+      mxnet::cpp::Operator("_random_uniform")(begin, end)
+          .Invoke((*data_arr)[i]);
+    }
+    mxnet::cpp::NDArray::WaitAll();
+  }
+}
+
+// Run inference on a model
+void run_inference(const std::string& model_name, const std::vector<mxnet::cpp::NDArray>& input_arrs,
```

Comment:
just mxnet::NDArray?

Reply:
the cpp ndarray is a frontend ndarray which is implemented like any other frontend ndarray i.e. save the ndarray handle (pointer to the ndarray) and perform C API operations on it.